### PR TITLE
Allow setting vterm ansi color palette in GUI color mode

### PR DESCRIPTION
**Pull request by andymass**
Problem: when using a GUI or with termguicolors, there is no way to configure the color palette used in the `:terminal` window.
Solution: allow the colors to be changed with the variables `g:terminal_color_0` through `g:terminal_color_15`.

As with the foreground and background colors which are set via `hi Terminal`, changes to these variables will only apply to newly created terminal windows.

To get the full 16 colors it's necessary to use `vterm_state_set_bold_highbright`, otherwise colors 8-15 appear the same as 0-7 (though the font is bolded).  I don't think this needs to be configurable.

Question: do the `term_dump*` functions need to be modified too?  I'm not familiar with them.

### CODE REVIEW

**Comment by chrisbra** [comment]
While I am sympathetic with keeping in sync with the Neovim change, I don't think setting internal Vim variables is a good way to specify the color. Isn't it better to have some kind of option or perhaps use a highlighting group to set those colors? That comes with the advantage that the user can enter `:verbose ...` and check where that was actually set, while simply setting a variable leaves no clue which script set those.

**Comment by brammool** [comment]
Christian wrote:

> While I am sympathetic with keeping in sync with the Neovim change, I
> don't think setting internal Vim variables is a good way to specify
> the color. Isn't it better to have some kind of option or perhaps use
> a highlighting group to set those colors? That comes with the
> advantage that the user can enter `:verbose ...` and check where that
> was actually set, while simply setting a variable leaves no clue which
> script set those.

It should at least be one variable with a list instead of using 16
variables.

Also, the colors apply to a Vterm instance, they should not be global.
Might want to pass them in term_start() options.
And then we need a function to set them afterwards.

-- 
hundred-and-one symptoms of being an internet addict:
45. You buy a Captain Kirk chair with a built-in keyboard and mouse.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by vim-ml** [comment]
2018-03-24 8:41 GMT+01:00 Christian Brabandt <vim-dev-github@256bit.org>:
> While I am sympathetic with keeping in sync with the Neovim change, I don't
> think setting internal Vim variables is a good way to specify the color.
> Isn't it better to have some kind of option or perhaps use a highlighting
> group to set those colors? That comes with the advantage that the user can
> enter :verbose ... and check where that was actually set, while simply
> setting a variable leaves no clue which script set those.
>

The global variables are a hack, I suggest instead highlight group
names like TermColor1, TermColor2, for each of the color indices.

Example:

    :hi TermColor4 guifg=lightblue ctermfg=lightblue

---
Justin M. Keyes

**Comment by vim-ml** [comment]
Justin M. Keyes wrote:

> 2018-03-24 8:41 GMT+01:00 Christian Brabandt <vim-dev-github@256bit.org>:
> > While I am sympathetic with keeping in sync with the Neovim change, I don't
> > think setting internal Vim variables is a good way to specify the color.
> > Isn't it better to have some kind of option or perhaps use a highlighting
> > group to set those colors? That comes with the advantage that the user can
> > enter :verbose ... and check where that was actually set, while simply
> > setting a variable leaves no clue which script set those.
> >
> 
> The global variables are a hack, I suggest instead highlight group
> names like TermColor1, TermColor2, for each of the color indices.
> 
> Example:
> 
>     :hi TermColor4 guifg=lightblue ctermfg=lightblue

Hmm, although this could be made to work, there are some conceptual
differences between a highlight group and the ANSI colors of a
terminal.  Highlight groups are meant to be global, specify the colors
of some language or status line elements.  The colors of the terminal
might depend on what's running in that terminal.  And a highlight group
provides attributes, which would not apply here.


-- 
hundred-and-one symptoms of being an internet addict:
50. The last girl you picked up was only a jpeg.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by andymass** [comment]
I agree with all of you regarding `g:` variables.  The choices I see are are a single array global variable (somewhat hackish), highlight groups (conceptually weird), options (more difficult to add an option and decide on the semantics), or `term_setcolors`/`term_setpalette` (easiest implementation).  I would try the function option, if it did not logically conflict with the already implemented "Terminal" highlight group. 

> there are some conceptual differences between a highlight group and the ANSI colors of a terminal

The "Terminal" highlight group already has this problem.  Changing it only applies to newly created terminals, old ones don't change.  I think there is one advantage to highlight groups, that the attributes would allow you to specify whether bright colors are bolded or not.  But either having Term(inal)Color0-7 with guifg/bg for normal and bright or Term(inal)Color0-15 with guifg only taking effect are both strange.

Highlight groups would make more sense if terminals kept their coloring after closing, when becoming normal buffers.  Then, every terminal would have the same global colors, but they would all change immediately when changing the highlight groups.  I don't know if "terminals becoming colored buffers" is on the road-map, how difficult this would be to implement, or if it's even worth it.

**Comment by brammool** [comment]
Let's use a list with 16 entries for this.  A plugin (or the user) would create this once and give it a meaningful name.  Then when creating a terminal window it can be passed during creation.  And it can be changed afterwards with a function call.

The Terminal highlight group does something else, it doesn't set the ANSI colors, it uses them.

This is one of the last pieces of the terminal functionality planned to be implemented, I hope you can do this soon.  Otherwise I'll have a go at it.

**Comment by andymass** [comment]
Updated (tests needed still).  If I understand your requirements correctly, this adds the following:

 1. An option entry to `term_start` called `term_colors` which accepts a length 16 list of colors (names or hex codes).
 
 2. If the option is not specified (e.g., using `:terminal`), colors are taken from the list variable `g:terminal_colors` in the same format.
 
3. A function `term_setguicolors(buf, colors)` changes the colors after creation in the same format.

Should (and can) the terminal window be refreshed?  Also, names of everything are up for debate.

**Comment by brammool** [comment]
> Updated (tests needed still).  If I understand your requirements correctly, this adds the following:
> 
>  1. An option entry to `term_start` called `term_colors` which accepts a length 16 list of colors (names or hex codes).

"ansi_colors" is a better name.

>  2. If the option is not specified (e.g., using `:terminal`), colors are taken from the list variable `g:terminal_colors` in the same format.

g:terminal_ansi_colors

> 3. A function `term_setguicolors(buf, colors)` changes the colors after creation in the same format.

term_setansicolors()

We probably also want term_getansicolors() (if a plugin wants to figure
out what colors are actually being used).

> Should (and can) the terminal window be refreshed?  Also, names of everything are up for debate.

Doesn't libvterm already trigger a redraw when setting colors?
Anyway, calling redraw_later() with NOT_VALID should probably be used.

-- 
hundred-and-one symptoms of being an internet addict:
114. You are counting items, you go "0,1,2,3,4,5,6,7,8,9,A,B,C,D...".

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by andymass** [comment]
> We probably also want term_getansicolors()

Would that be the effective RGB color values, or the actual strings passed by the user ('LightBlue', etc) which should then be stored somewhere?

> Doesn't libvterm already trigger a redraw when setting colors?

No, after calling `term_setansicolors`, new terminal output has the new colors but old terminal output still has the old colors.

**Comment by brammool** [comment]
> > We probably also want term_getansicolors()
> 
> Would that be the effective RGB color values, or the actual strings passed by the user ('LightBlue', etc) which should then be stored somewhere?

Good questions.  If someone wants to save the values to restore later, I
suppose both should work.  If someone wants to figure out what's being
displayed, then the RGB values are probably needed.  So the RGB values
might work in both cases.

> > Doesn't libvterm already trigger a redraw when setting colors?
> 
> No, after calling `term_setansicolors`, new terminal output has the new colors but old terminal output still has the old colors. 

Then preparing to redraw later should be done by Vim.

-- 
hundred-and-one symptoms of being an internet addict:
116. You are living with your boyfriend who networks your respective
     computers so you can sit in separate rooms and email each other

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by andymass** [comment]
Updated with a test.  Unfortunately, `redraw_later(NOT_VALID)`/`redraw_buf_later(buf, NOT_VALID)` don't seem to work.  Old text in the terminal still stays with the same colors.

File: runtime/doc/eval.txt [hunk 1]
```diff
@@ -8248,6 +8251,11 @@ term_getaltscreen({buf})				*term_getaltscreen()*
 		{buf} is used as with |term_getsize()|.
 		{only available when compiled with the |+terminal| feature}
 
+term_getansicolors({buf})				*term_getansicolors()*
+		Get the ANSI color palette in use by terminal {buf}.
+		Returns a List of length 16 where each element is a String
+		representing a color in "#rrggbb" format.
```
Comment (chdiza) [inline runtime/doc/eval.txt]: If {buf} is not given, it should default to the current buffer.  That needs to be noted here in the docs.

Comment (andymass) [reply runtime/doc/eval.txt]: Thanks, documented.  To be consistent with the other `term_` functions, `term_getansicolors('')` refers to the current buffer (`term_getansicolors('%')` also works).

**Comment by brammool** [comment]
Andy Massimino wrote:

> Updated with a test.  Unfortunately,
> `redraw_later(NOT_VALID)`/`redraw_buf_later(buf, NOT_VALID)` don't
> seem to work.  Old text in the terminal still stays with the same
> colors.

Looking at vterm_state_set_palette_color(), it just stores the new
color.  Thus all previously written text won't change.

I can understand why this happens, but it's not really what the user
expects.  If one changes the color palette, one expects it to apply to
existing text as well.

It's going to be really difficult to change the behavior.  Since I would
expect the colors to be set once at the start and not change later, I
don't think it is worth the effort to change it.

-- 
I have a watch cat! Just break in and she'll watch.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by bluz71** [comment]
So the agreed upon syntax is a single variable `g:terminal_ansi_colors` which takes a list of 16 colors?

Is the following a valid usage of the new variable:

```
    let g:terminal_ansi_colors = [
                \ "#373c40", "#ff5454", "#8cc85f", "#e3c78a",
                \ "#80a0ff", "#ce76e8", "#7ee0ce", "#de935f",
                \ "#f09479", "#f74782", "#42cf89", "#cfcfb0",
                \ "#78c2ff", "#ae81ff", "#85dc85", "#e2637f"
                \]
```

P.S Thanks for doing this, it will be a very good addition and will very much help us Vim theme authors.

**Comment by machakann** [comment]
I'm not so sure the expected usage, but if it is expected to be set in a colorscheme file, wouldn't it be better to clear `g:terminal_ansi_colors` in `syntax/nosyntax.vim`? Otherwise the old color set may still take effect even after changing a color theme.

File: runtime/doc/eval.txt (hunk 1 above)
Comment (egberts) [reply runtime/doc/eval.txt]: These functions (particularly `term_getansicolors()`) are currently returned as `[]`.  Is this the expected value (`[]`) at startup?

```bash
vim -g --clean
```
then
```vim
:verbose echo term_getansicolors('')
[]
" Details
:verbose echo &term has("gui_running") &t_Co
builtin_gui 1  
```

## Details ##
VIM - Vi IMproved 8.2 (2019 Dec 12, compiled Apr 10 2020 09:45:15)
Included patches: 1-539
Huge version with GTK2 GUI.

**Pull request closed without merging**